docs/proposals/cluster-discovery-api.md

```diff
@@ -0,0 +1,91 @@
+# Cluster Discovery API
+
+Author: [@mikedanese](https://github.com/mikedanese)
+
+#### Abstract
+
+This document describes a mechanism for a kubelet to discover, from a trusted source, the
+location of the cluster API and the public key of the cluster's local root certificate
+authority.
+
+#### Motivation
+
+Setting up a kubernetes cluster is hard. Setting up a secure kubernetes cluster is harder.
+Public key infrastructure is an area that most of our users are unfamiliar with and may
+not care about for many cluster deployments where network securityis not a concern. The
+Kubernetes Discovery API provides a mechanism to enable a high level of security by
+default without compromising on the UX of the cluster deployment flow.
+
+#### Overview
+
+The Kubelet TLS Bootstrap API reduces the amount of manual steps required to generate and
+distribute cryptographic assets required to setup a secure Kubernetes cluster. Still, in
+order to initiate the kubelet TLS bootstrap process, the user is required to distribute
+the URLs of the api-server and the root CA public key to each node. Today, there is no
+mechanism built into Kubernetes to facilitate this.
+
+We can reduce this configuration to a single 256 bit cluster-id that can be easily passed
+into the commandline of the kubelet by hand or otherwise. This cluster-id would be used to
+lookup a discovery.Cluster object (documented bellow) in a globally addressable and trusted
+service (by default https://discovery.k8s.io, but also internally hostable). The default
+service's certificate will be signed by a CA whose certificate is packaged by the OS
+distribution (i.e a public and trusted CA).
+
+On first boot, kubelet would accept the cluster-id as configuration and use it to lookup the
+discovery.Cluster object. The discovery.Cluster object would have the required information to
+establish a trusted first contact with the cluster api and initiate the kubelet TLS bootstrap
+flow.
+
+#### API Types
+
+Hosted at:
+
+```
+PUT https://${DISCOVERY_SERVICE}/v1alpha1/clusters
+GET https://${DISCOVERY_SERVICE}/v1alpha1/clusters/<cluster-id>
+```
+
+A user can create and retrieve the cluster object:
+
+```
+type Cluster struct {
+	unversioned.TypeMeta `json:",inline"`
+	api.ObjectMeta       `json:"metadata,omitempty"`
+	RootCAPublicKey      []byte   `json:"rootCAPublicKey"`
+	ApiServerURLs        []string `json:"apiServerURLs"`
+}
+```
+
+Cluster objects are immutable after creation.
+
+#### Validating the Integrity of the discovery.Cluster object
+
+We can make the Cluster object self validating if the cluster-id is generated from a consistent
+hashing of a subset of the object. For example the cluster-id colud be generated from a sha256
+sum of an object containing the fields `.type`, `.apiVersion`, `.metadata.annotations`,
+`.rootCAPublicKey`, `.apiServerURLs` serialized to JSON. The administrator or automation tooling
```

Comment:
Per above question from @luxas, this is probably why it was documented that Cluster objects were immutable. However it feels like that list of API Servers and probably the root CA public key may both need to be able to change, without obsoleting the token.